### DISCUSSION TITLE
Should we stop overloading NumPy altogether?

### DISCUSSION THREAD

**Discussion by ricardoV94** [aesara-devs/aesara#802] [General]
Overloading NumPy functions when the input are TensorVariables is cute, but I think confusing as soon as a user tries to use a function that is not implemented or requires a different API (like set_subtensor)

From what I see around it seems that knowledgeable users don't rely on this at all and it's beginners who use it, usually by accident. I think it's more harmful than helpful.

```python
import numpy as np
import aesara.tensor as at

x = at.scalar('x')
y = np.exp(x)
y  # Elemwise{exp,no_inplace}.0
```

**Comment by twiecki** [comment]
I do some benefit, it mostly works with just simple functions with `np.exp()` and new-comers seem to do this pretty automatically and are not confused when it works.

**Comment by brandonwillard** [comment]
We need to do a lot more to make it _actually_ work.  There's a very narrow space where it's usable, and, due to that, the feature is **more** confusing than not.  

A good example is `np.dot`; I've accidentally used that a few times instead of `at.dot` and it led to extremely confusing errors&mdash;even by my standards.

**Comment by dhirschfeld** [comment]
It can be convenient for an array implementation to Just Work with the numpy methods but if it sometimes does the wrong thing that would be very bad.

IIUC you can implement [`__array_function__`](https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_function__) to specify what you support, or return `NotImplemented` to disable the dispatching altogether.

**Reply by brandonwillard** [reply]
> It can be convenient for an array implementation to Just Work with the numpy methods but if it sometimes does the wrong thing that would be very bad.

That's exactly the case, so I'm all for us removing the partial support until someone does it correctly with `__array_function__` or whatnot.
